exercises/forth/example.go

```diff
@@ -0,0 +1,243 @@
+		func(r rune) bool {
+			return unicode.IsSpace(r) || unicode.IsControl(r)
+		})
+
+	var t int // t is token index into words[]
+	for t = 0; t < len(words); t++ {
+		w := strings.ToUpper(words[t])
+		// Handle reference to user defined word.
+		if udef, ok := userDefs[w]; ok {
+			oplist = append(oplist, udef...)
+		} else if op, ok := builtinOps[w]; ok {
+			if op.id == userDefOp {
+				// Handle user defined word definition.
+				t++
+				if t >= len(words)-2 {
+					return nil, emptyUserDefErr
+				}
+				userword := strings.ToUpper(words[t])
+				_, numerr := strconv.Atoi(userword)
+				if numerr == nil {
+					return nil, invalidUserDefErr
+				}
+				t++
+				var userops []operatorTyp
+				for t < len(words) {
+					oneOp, err := parse(words[t], userDefs)
+					if err != nil {
+						return nil, err
+					}
+					if oneOp[0].id == endDefOp {
+						break
+					}
+					userops = append(userops, oneOp...)
+					t++
+				}
+				if len(userops) == 0 {
+					return nil, emptyUserDefErr
+				} else {
+					userDefs[userword] = userops
+				}
+			} else {
+				// Normal builtin operator.
+				oplist = append(oplist, op)
+			}
+		} else {
+			// Handle constant literal.
+			var x int
+			x, err = strconv.Atoi(w)
+			if err != nil {
+				return nil, err
+			}
+			oplist = append(oplist,
+				operatorTyp{id: constOp,
+					fn: func(stack *[]int) error {
+						push(stack, x)
+						return nil
+					},
+				})
+		}
+	}
+	return oplist, nil
+}
+
+// builtinOps are the pre-defined operators to support.
+var builtinOps = map[string]operatorTyp{
+	"+":    {add, addOp},
+	"-":    {subtract, subOp},
+	"*":    {multiply, mulOp},
+	"/":    {divide, divOp},
+	"DUP":  {dup, dropOp},
+	"DROP": {drop, dupOp},
+	"SWAP": {swap, swapOp},
+	"OVER": {over, overOp},
+	":":    {nil, userDefOp},
+	";":    {nil, endDefOp},
+}
+
+func pop(stack *[]int) (v int, err error) {
+	slen := len(*stack)
+	if slen >= 1 {
+		v = (*stack)[slen-1]
+		*stack = (*stack)[:slen-1]
+		return v, nil
+	}
+	return 0, notEnoughOperands
+}
+
+func pop2(stack *[]int) (v1, v2 int, err error) {
+	v1, err = pop(stack)
+	if err != nil {
+		return 0, 0, notEnoughOperands
+	}
+	v2, err = pop(stack)
+	return v1, v2, err
+}
+
+func push(stack *[]int, v int) {
+	*stack = append(*stack, v)
+}
+
+func add(stack *[]int) (err error) {
+	v1, v2, err := pop2(stack)
+	if err != nil {
+		return err
+	}
+	push(stack, v1+v2)
+	return nil
+}
+
+func subtract(stack *[]int) error {
+	v1, v2, err := pop2(stack)
+	if err != nil {
+		return err
+	}
+	push(stack, v2-v1)
+	return nil
+}
+
+func multiply(stack *[]int) error {
+	v1, v2, err := pop2(stack)
+	if err != nil {
+		return err
+	}
+	push(stack, v2*v1)
+	return nil
+}
+
+func divide(stack *[]int) error {
+	v1, v2, err := pop2(stack)
+	if err != nil {
+		return err
+	}
+	if v1 == 0 {
+		return divideByZero
+	}
+	push(stack, v2/v1)
+	return nil
+}
+
+func dup(stack *[]int) error {
+	v1, e1 := pop(stack)
+	if e1 != nil {
+		return notEnoughOperands
```

Comment:
hmm, question, why unconditionally return `notEnoughOperands` here, rather than `e1`? What if `e1` were a different kind of error?

Then again, we know that `pop` will only ever return `notEnoughOperands`, so this question is only academic. So I guess the only benefit to saying `return e1` would be "it saves you the extra typing".